File: lc-python/src/lc_0110_balanced_binary_tree.py
import unittest
from typing import List, Optional

from src.shared.tree import TreeNode, list_tree_to_tree


def is_balanced(root: Optional[TreeNode]) -> bool:
    def bfs(node: Optional[TreeNode]) -> int:
        if node is None:
            return 0
        dl = bfs(node.left)
        dr = bfs(node.right)
        if dl == -1 or dr == -1 or abs(dl - dr) > 1:
            return -1
        return 1 + max(dl, dr)

    return bfs(root) != -1


class TestBalancedBinaryTree(unittest.TestCase):
    def test_balanced_binary_tree(self):
        test_cases = [
            [[3, 9, 20, None, None, 15, 7], True],
            [[1, 2, 2, 3, 3, None, None, 4, 4], False],
            [[], True],
        ]
        for i, tc in enumerate(test_cases):
            t = list_tree_to_tree(tc[0])
            actual = is_balanced(t)
            expected = tc[1]
            self.assertEqual(actual, expected, f"{i} {tc}")


if __name__ == "__main__":
    unittest.main()

File: lc-typescript/src/lc_0110_balanced_binary_tree.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/assert_equals.ts";
import { array_to_tree, TreeNode } from "./shared/utils.ts";

function is_balanced1(root: TreeNode | null): boolean {
    function dfs(node: TreeNode | null): number {
        if (node == null) {
            return 0;
        }

        const left = dfs(node.left);
        if (left < 0) {
            return -1;
        }

        const right = dfs(node.right);
        if (right < 0) {
            return -1;
        }

        const diff = left > right ? left - right : right - left;
        if (diff > 1) {
            return -1;
        }

        return 1 + Math.max(left, right);
    }

    return dfs(root) != -1;
}

function is_balanced2(root: TreeNode | null): boolean {
    function dfs(node: TreeNode | null): number {
        if (!node) {
            return 0;
        }
        const lhs = dfs(node.left);
        const rhs = dfs(node.right);
        const diff = Math.abs(lhs - rhs);
        if (lhs < 0 || rhs < 0 || diff > 1) {
            return -1;
        }
        return 1 + Math.max(lhs, rhs);
    }

    return dfs(root) !== -1;
}

Deno.test("0110 Balanced Binary Tree", () => {
    const test_cases: [(number | null)[], boolean][] = [
        [[3, 9, 20, null, null, 15, 7], true],
        [[1, 2, 2, 3, 3, null, null, 4, 4], false],
        [[], true],
    ];

    for (let i = 0; i < test_cases.length; i += 1) {
        const test_case = test_cases[i];

        const root = array_to_tree(test_case[0]);

        const expected = test_case[1];

        const actual1 = is_balanced1(root);
        assertEquals(actual1, expected);

        const actual2 = is_balanced2(root);
        assertEquals(actual2, expected);
    }
});
